extension/healthcheckv2extension/internal/status/aggregator.go

```diff
@@ -0,0 +1,225 @@
+// whether or not the scope was found.
+func (a *Aggregator) AggregateStatus(scope Scope, verbosity Verbosity) (*AggregateStatus, bool) {
+	a.mu.RLock()
+	defer a.mu.RUnlock()
+
+	if scope == ScopeAll {
+		return a.aggregateStatus.clone(verbosity), true
+	}
+
+	st, ok := a.aggregateStatus.ComponentStatusMap[scope.toKey()]
+	if !ok {
+		return nil, false
+	}
+
+	return st.clone(verbosity), true
+}
+
+// RecordStatus stores and aggregates a StatusEvent for the given component instance.
+func (a *Aggregator) RecordStatus(source *component.InstanceID, event *component.StatusEvent) {
+	compIDs := source.PipelineIDs
+	// extensions are treated as a pseudo-pipeline
+	if source.Kind == component.KindExtension {
+		compIDs = extsIDMap
+	}
+
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	for compID := range compIDs {
+		var pipelineStatus *AggregateStatus
+		pipelineScope := Scope(compID.String())
+		pipelineKey := pipelineScope.toKey()
+
+		pipelineStatus, ok := a.aggregateStatus.ComponentStatusMap[pipelineKey]
+		if !ok {
+			pipelineStatus = &AggregateStatus{
+				ComponentStatusMap: make(map[string]*AggregateStatus),
+			}
+		}
+
+		componentKey := fmt.Sprintf("%s:%s", strings.ToLower(source.Kind.String()), source.ID)
+		pipelineStatus.ComponentStatusMap[componentKey] = &AggregateStatus{
+			Event: event,
+		}
+		a.aggregateStatus.ComponentStatusMap[pipelineKey] = pipelineStatus
+		pipelineStatus.Event = a.aggregationFunc(pipelineStatus)
+		a.notifySubscribers(pipelineScope, pipelineStatus)
+	}
+
+	a.aggregateStatus.Event = a.aggregationFunc(a.aggregateStatus)
+	a.notifySubscribers(ScopeAll, a.aggregateStatus)
+}
+
+// Subscribe allows you to subscribe to a stream of events for the given scope. The scope can be
+// the collector overall (ScopeAll), extensions (ScopeExtensions), or a pipeline name.
+// It is possible to subscribe to a pipeline that has not yet reported. An initial nil
+// will be sent on the channel and events will start streaming if and when it starts reporting.
+// Detail specifies whether or not subtrees should be returned with the *AggregateStatus.
+func (a *Aggregator) Subscribe(scope Scope, verbosity Verbosity) <-chan *AggregateStatus {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	key := scope.toKey()
+	st := a.aggregateStatus
+	if scope != ScopeAll {
+		st = st.ComponentStatusMap[key]
+	}
+	if st != nil {
+		st = st.clone(verbosity)
+	}
+	sub := &subscription{
+		statusCh:  make(chan *AggregateStatus, 1),
+		verbosity: verbosity,
+	}
+
+	a.subscriptions[key] = append(a.subscriptions[key], sub)
+	sub.statusCh <- st
+
+	return sub.statusCh
+}
+
+// Unbsubscribe removes a stream from further status updates.
+func (a *Aggregator) Unsubscribe(statusCh <-chan *AggregateStatus) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	for scope, subs := range a.subscriptions {
+		for i, sub := range subs {
+			if sub.statusCh == statusCh {
+				a.subscriptions[scope] = append(subs[:i], subs[i+1:]...)
```

Comment:
Not totally sure what the efficiency of this is, but we may want to consider using a doubly-linked list if we expect a large number of subscribers.

The OpAMP extension [implements a similar notification pattern](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/extension/opampextension/registry.go#L28) and uses the Go stdlib `container/list` package to try to make removals more efficient.

Reply:
I started down the road of using a list and it became uglier than I had hoped. When subscribing the caller gets a `chan <- *AggregateStatus`, and this is what is used to unsubscribe. Internally, the subscription is a struct with a channel and verbosity. I'd prefer not to return that to the caller, because only the channel is of any use. The net of this, is that you need to iterate through all of the subscriptions to find one with the matching channel. In order to make this work, I'd have to introduce two lookup maps. One that maps the channel to its subscription, and one that maps the subscription to the list containing it. It can be done, but this might be a premature optimization. We can certainly do this if it's needed in the future. Is that reasonable or am I missing something that makes this easier?

Reply:
For additional context, this functionality is used for gRPC streaming healthchecks only. I do not expect an overly large number of clients.

Reply:
I took a second look at this and updated subscribe to return a channel and a function to unsubscribe. The UnsubscribeFunc cleared up the issues I was running into and this now uses container/list for subscriptions.